### PR TITLE
feat: add chain processor implementation

### PR DESCRIPTION
**Pull request by TxCorpi0x**
## Description
Chain processor implementation is mainly cloned from cosmos implementation and manipulated according to the substrate requirements and specifications. 

## Implementations
- [x] Implement `latestHeightWithRetry`
- [x] Implement `clientState`
- [x] Implement `Run`
- [x] Implement `initializeConnectionState`
- [x] Implement `initializeChannelState`
- [x] Implement  `queryCycle` 

## Considerations
In a substrate chain, we can not query blocks individually, So we need to query an RPC endpoint of the composable chain to get un-relayed events.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -34,11 +33,7 @@ require (
 	filippo.io/edwards25519 v1.0.0-rc.1 // indirect
 	github.com/99designs/go-keychain v0.0.0-20191008050251-8e49817e8af4 // indirect
 	github.com/99designs/keyring v1.2.1 // indirect
-	github.com/Azure/go-ansiterm v0.0.0-20210617225240-d185dfc1b5a1 // indirect
-	github.com/ChainSafe/go-schnorrkel v0.0.0-20200405005733-88cbf1b4c40d // indirect
-	github.com/Microsoft/go-winio v0.5.2 // indirect
-	github.com/Nvveen/Gotty v0.0.0-20120604004816-cd527374f1e5 // indirect
-	github.com/Workiva/go-datastructures v1.0.53 // indirect
+	github.com/ChainSafe/go-schnorrkel v1.0.0 // indirect
```
Comment (agouin) [inline go.mod]: Same as https://github.com/cosmos/relayer/pull/923#discussion_r949328783

Comment (TxCorpi0x) [reply go.mod]: Has been fixed

**Comment by jtieri** [comment]
@vjdmhd can we get the merge conflicts solved here + tests passing before review is submitted?

**Comment by TxCorpi0x** [comment]
> @vjdmhd can we get the merge conflicts solved here + tests passing before review is submitted?

Hey @jtieri the feat-substrate is merged to the current branch, can you kindly run the workflows to check?

File: relayer/chains/substrate/substrate_chain_processor.go [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+		})
+
+		if err := eg.Wait(); err != nil {
+			scp.log.Warn("Error querying block data", zap.Error(err))
+			break
+		}
+
+		latestHeader = ibcHeader.(SubstrateIBCHeader)
+
+		// TODO: update latest bolock
```
Comment (jtieri) [inline relayer/chains/substrate/substrate_chain_processor.go:323]: Should this still be commented out? After comparing the Substrate processor with the Cosmos processor, it looks like this is the only piece missing.

**Review by jtieri — APPROVED** [review]
Double checked everything and this LGTM. I had one question regarding updating the latest block but I'm sure you have more context on this than me so perhaps there is something else missing here before this can be done.

File: relayer/chains/substrate/substrate_chain_processor.go (hunk 2 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/substrate_chain_processor.go:323]: Done in the [PR#986](https://github.com/cosmos/relayer/pull/986)

**Pull request merged**